File: app/src/main/java/com/platdmit/simplecloudmanager/data/api/implement/ApiDomainRepoImp.java
package com.platdmit.simplecloudmanager.data.api.implement;

import com.platdmit.simplecloudmanager.data.api.ApiDomainRepo;
import com.platdmit.simplecloudmanager.data.api.ApiManager;
import com.platdmit.simplecloudmanager.data.api.models.ApiDomain;
import com.platdmit.simplecloudmanager.data.api.models.ApiDomainRecord;
import com.platdmit.simplecloudmanager.data.api.models.ApiRequestBody;
import com.platdmit.simplecloudmanager.data.api.rest.RestDomain;
import com.platdmit.simplecloudmanager.domain.helpers.ActualApiKeyService;

import java.util.List;

import io.reactivex.rxjava3.core.Single;
import retrofit2.Response;

public class ApiDomainRepoImp implements ApiDomainRepo {
    private static final String TAG = ApiDomainRepoImp.class.getSimpleName();
    private ActualApiKeyService mActualApiKeyService;

    public ApiDomainRepoImp(ActualApiKeyService actualApiKeyService) {
        mActualApiKeyService = actualApiKeyService;
    }

    private RestDomain getRestDomain(){
        //always gets the actual key
        return ApiManager.<RestDomain>getApiPoint(RestDomain.class, mActualApiKeyService.getApiKey());
    }

    @Override
    public Single<List<ApiDomain>> getDomains() {
        return Single.create(requestResult -> {
            try {
                Response<ApiRequestBody> response = getRestDomain().getDomains().execute();
                if (response.isSuccessful()){
                    ApiRequestBody apiRequestBody = response.body();
                    requestResult.onSuccess(apiRequestBody.getDomains());
                } else {
                    throw new Throwable(response.message());
                }
            } catch (Exception e) {
                requestResult.onError(e);
            }
        });
    }

    @Override
    public Single<ApiDomain> getDomain(long domainId) {
        return Single.create(requestResult -> {
            try {
                Response<ApiRequestBody> response = getRestDomain().getDomain(domainId).execute();
                if (response.isSuccessful()){
                    ApiRequestBody apiRequestBody = response.body();
                    requestResult.onSuccess(apiRequestBody.getDomain());
                } else {
                    throw new Throwable(response.message());
                }
            } catch (Exception e) {
                requestResult.onError(e);
            }
        });
    }

    @Override
    public Single<List<ApiDomainRecord>> getDomainRecords(long domainId) {
        return Single.create(requestResult -> {
            try {
                Response<ApiRequestBody> response = getRestDomain().getDomainRecords(domainId).execute();
                if (response.isSuccessful()){
                    ApiRequestBody apiRequestBody = response.body();
                    requestResult.onSuccess(apiRequestBody.getDomainRecords());
                } else {
                    throw new Throwable(response.message());
                }
            } catch (Exception e) {
                requestResult.onError(e);
            }
        });
    }

    @Override
    public Single<ApiDomainRecord> getDomainRecord(int domainId, int recordId) {
        return Single.create(requestResult -> {
            try {
                Response<ApiRequestBody> response = getRestDomain().getDomain(domainId).execute();
                if (response.isSuccessful()){
                    ApiRequestBody apiRequestBody = response.body();
                    requestResult.onSuccess(apiRequestBody.getDomainRecord());
                } else {
                    throw new Throwable(response.message());
                }
            } catch (Exception e) {
                requestResult.onError(e);
            }
        });
    }
}

File: data/src/main/java/com/platdmit/data/api/implement/ApiDomainRepoImp.kt
package com.platdmit.data.api.implement

import com.platdmit.data.api.ApiDomainRepo
import com.platdmit.data.api.ApiManager
import com.platdmit.data.api.models.ApiDomain
import com.platdmit.data.api.models.ApiDomainRecord
import com.platdmit.data.api.rest.RestDomain
import com.platdmit.data.helpers.ActualApiKeyService
import io.reactivex.rxjava3.core.Single
import java.io.IOException

class ApiDomainRepoImp(private val mActualApiKeyService: ActualApiKeyService) : ApiDomainRepo {
    //always gets the actual key
    private val restDomain: RestDomain
        get() = ApiManager.getApiPoint(RestDomain::class.java, mActualApiKeyService.apiKey)

    override fun getDomains(): Single<List<ApiDomain>> {
        return Single.create {
            try {
                val response = restDomain.getDomains().execute()
                if (response.isSuccessful) {
                    val apiRequestBody = response.body()
                    it.onSuccess(apiRequestBody!!.domains)
                } else {
                    throw Throwable(response.message())
                }
            } catch (e: IOException) {
                println(TAG + e.localizedMessage)
            } catch (e: Exception) {
                println(TAG + e.localizedMessage)
                it.onError(e)
            }
        }
    }

    override fun getDomain(domainId: Long): Single<ApiDomain> {
        return Single.create {
            try {
                val response = restDomain.getDomain(domainId).execute()
                if (response.isSuccessful) {
                    val apiRequestBody = response.body()
                    it.onSuccess(apiRequestBody!!.domain)
                } else {
                    throw Throwable(response.message())
                }
            } catch (e: IOException) {
                println(TAG + e.localizedMessage)
            } catch (e: Exception) {
                println(TAG + e.localizedMessage)
                it.onError(e)
            }
        }
    }

    override fun getDomainRecords(domainId: Long): Single<List<ApiDomainRecord>> {
        return Single.create {
            try {
                val response = restDomain.getDomainRecords(domainId).execute()
                if (response.isSuccessful) {
                    val apiRequestBody = response.body()
                    it.onSuccess(apiRequestBody!!.domainRecords)
                } else {
                    throw Throwable(response.message())
                }
            } catch (e: IOException) {
                println(TAG + e.localizedMessage)
            } catch (e: Exception) {
                println(TAG + e.localizedMessage)
                it.onError(e)
            }
        }
    }

    override fun getDomainRecord(domainId: Int, recordId: Int): Single<ApiDomainRecord> {
        return Single.create {
            try {
                val response = restDomain.getDomain(domainId.toLong()).execute()
                if (response.isSuccessful) {
                    val apiRequestBody = response.body()
                    it.onSuccess(apiRequestBody!!.domainRecord)
                } else {
                    throw Throwable(response.message())
                }
            } catch (e: IOException) {
                println(TAG + e.localizedMessage)
            } catch (e: Exception) {
                println(TAG + e.localizedMessage)
                it.onError(e)
            }
        }
    }

    companion object {
        private val TAG = ApiDomainRepoImp::class.java.simpleName
    }

}